component/s3/file/s3.go

```diff
@@ -0,0 +1,223 @@
+package s3
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	aws_config "github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/flow/rivertypes"
+	"github.com/grafana/agent/pkg/river"
+	"golang.org/x/net/context"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "s3.file",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// S3 handles reading content from a file located in S3
+type S3 struct {
+	mut        sync.Mutex
+	opts       component.Options
+	args       Arguments
+	health     *component.Health
+	watcher    *watcher
+	updateChan chan []byte
+	errorChan  chan error
+	content    string
+	cancel     context.CancelFunc
+}
+
+var (
+	_ component.Component       = (*S3)(nil)
+	_ component.HealthComponent = (*S3)(nil)
+	_ river.Unmarshaler         = (*Arguments)(nil)
+)
+
+// New initializes the s3 component
+func New(o component.Options, args Arguments) (*S3, error) {
+	s3cfg, err := generateS3Config(args)
+	if err != nil {
+		return nil, err
+	}
+
+	s3Client := s3.NewFromConfig(*s3cfg, func(s3o *s3.Options) {
+		s3o.UsePathStyle = args.Options.UsePathStyle
+	})
+
+	bucket, file := getPathBucketAndFile(args.Path)
+	s := &S3{
+		opts:       o,
+		args:       args,
+		health:     &component.Health{},
+		updateChan: make(chan []byte),
+		errorChan:  make(chan error),
+	}
+
+	w, err := newWatcher(bucket, file, s.updateChan, s.errorChan, args.PollFrequency, s3Client)
+	if err != nil {
+		return nil, err
+	}
+	s.watcher = w
+	return s, nil
+}
+
+// Run activates the content handler and watcher
+func (s *S3) Run(ctx context.Context) error {
+	go s.handleContentUpdate(ctx)
+	go s.watcher.run(ctx)
+	<-ctx.Done()
+	if s.cancel != nil {
+		s.cancel()
+	}
+
+	return nil
+}
+
+// Update is called whenever the arguments have changed
+func (s *S3) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+	if newArgs.PollFrequency <= time.Second*30 {
```

Comment:
Randomly chosen minimum.

Reply:
IMO validations should be done at the Arguments level during unmarshal so it fails fast. 

(Also I'd make the minimum a global constant since it's a magic number)